lib/tasks/rubocop.rake

```diff
@@ -0,0 +1,45 @@
+#--
+# Yast rake
+#
+# Copyright (C) 2018 SUSE LLC
+#   This library is free software; you can redistribute it and/or modify
+# it only under the terms of version 2.1 of the GNU Lesser General Public
+# License as published by the Free Software Foundation.
+#
+#   This library is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+# details.
+#
+#++
+
+# run Rubocop in parallel
+# @param params [String] optional Rubocop parameters
+def run_rubocop(params = "")
+  # how it works:
+  # 1) get the list of inspected files by Rubocop
+  # 2) shuffle it randomly (better would be evenly distribute them according to
+  #    the analysis complexity but that is hard to evaluate and even simply
+  #    distributing by file size turned out to be ineffective and slower than
+  #    a simple random shuffling)
+  # 3) pass that as input for xargs
+  #    a) use -P with number of processors to run the commands in parallel
+  #    b) use -n to set the maximum number of files per process, this number
+  #       is computed to equally distribute the files across the workers
+  sh "rubocop -L | sort -R | xargs -P`nproc` -n$(expr `rubocop -L | wc -l` / " \
+    "`nproc` + 1) rubocop #{params}"
```

Comment:
this line is a bit hard to read. Why not precompute something in advance and maybe even write it in verbose mode? like number of processes? And why use only number of processors? Do you measure it? I expect part of rubocop will be also blocking IO for file reading, so maybe it make sense to use something like n+1 or 2n for number of processes. Can you measure it?

Reply:
I'll try making it more readable and verbose, that's a good idea.

Rubocop uses IO quite a lot for reading all files, so using more processes might help.

On the other hands, `nproc` already includes also the virtual processors (if hyperthreading is available and enabled). My quad core CPU runs 8 jobs, so it's already `2*n`, I'm not sure if using `2*(n+1)` would not actually decrease the speed. I'll do some testing...

Reply:
OK, some numbers for a quad core CPU with 8 threads (so `n=8` by default):

- `n = 9`: 12s
- `n = 6`: 12.5s
- `n = 4`: 14.6s 

So `n+1` would actually make it slightly slower. The question is how that would run on a CPU without HT, there `n+1` will probably help, the question is how much.

Also the question is if the complication is worth of trying that. I'd rather keep it simply unless the change would really make a difference.